engine/src/main/java/org/pentaho/di/trans/steps/formula/FormulaMeta.java

```diff
@@ -0,0 +1,240 @@
+
+package org.pentaho.di.trans.steps.formula;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.pentaho.di.core.CheckResult;
+import org.pentaho.di.core.CheckResultInterface;
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.core.database.DatabaseMeta;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleStepException;
+import org.pentaho.di.core.exception.KettleXMLException;
+import org.pentaho.di.core.row.RowMetaInterface;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.core.row.value.ValueMetaFactory;
+import org.pentaho.di.core.variables.VariableSpace;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.repository.ObjectId;
+import org.pentaho.di.repository.Repository;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.BaseStepMeta;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.step.StepMetaInterface;
+import org.pentaho.metastore.api.IMetaStore;
+import org.w3c.dom.Node;
+
+/**
+ * Contains the meta-data for the Formula step: calculates ad-hoc formula's Powered by Pentaho's "libformula"
+ *
+ * Created on 22-feb-2007
+ */
+
+public class FormulaMeta extends BaseStepMeta implements StepMetaInterface {
+  private static Class<?> PKG = FormulaMeta.class; // for i18n purposes, needed by Translator2!!
+
+  /** The formula calculations to be performed */
+  private FormulaMetaFunction[] formula;
+
+  public FormulaMeta() {
+    super(); // allocate BaseStepMeta
+  }
+
+  public FormulaMetaFunction[] getFormula() {
+    return formula;
+  }
+
+  public void setFormula( FormulaMetaFunction[] calcTypes ) {
+    this.formula = calcTypes;
+  }
+
+  public void allocate( int nrCalcs ) {
+    formula = new FormulaMetaFunction[nrCalcs];
+  }
+
+  public void loadXML( Node stepnode, List<DatabaseMeta> databases, IMetaStore metaStore ) throws KettleXMLException {
+    int nrCalcs = XMLHandler.countNodes( stepnode, FormulaMetaFunction.XML_TAG );
+    allocate( nrCalcs );
+    for ( int i = 0; i < nrCalcs; i++ ) {
+      Node calcnode = XMLHandler.getSubNodeByNr( stepnode, FormulaMetaFunction.XML_TAG, i );
+      formula[i] = new FormulaMetaFunction( calcnode );
+    }
+  }
+
+  public String getXML() {
+    StringBuilder retval = new StringBuilder();
+
+    if ( formula != null ) {
+      for ( int i = 0; i < formula.length; i++ ) {
+        retval.append( "       " + formula[i].getXML() + Const.CR );
+      }
+    }
+
+    return retval.toString();
+  }
+
+  public boolean equals( Object obj ) {
+    if ( obj != null && ( obj.getClass().equals( this.getClass() ) ) ) {
+      FormulaMeta m = (FormulaMeta) obj;
+      return ( getXML() == m.getXML() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.